[PROPOSAL] - Access android (MTP) devices

gave92 opened the pull request:
Might help fix #779

This PR contains a proposal of how MTP devices like android phones could be accessed by Files UWP.

What works:
* Navigation inside the MTP device
* Pasting/Deleting/Renaming inside the MTP device
* Creating new ~files~ folders inside the MTP device.
Files can only be copied to, explorer does the same
* Opening files inside the MTP device

@duke7553 what do you think?

duke7553 commented:
@gave92 I'm impressed by the amount of work you put into this, and will look forward to learning more about this change tomorrow

gave92 commented:
**Description of code changes**
The main issue when accessing MTP devices is that windows does not map them to normal drives. Files inside such devices will not have a filesystem "path", which results in the impossibility of using functions such as `StorageFolder.GetFolderFromPathAsync()` and `FindFirstFileExFromApp`.

The only way to get the contents of a folder is querying the StorageFolder with the `folder.GetFilesAsync()`/`folder.GetFoldersAsync()` methods.
Note: getting the parent of a StorageItem *also* doesn't work (`storageItem.GetParentAsync()` returns null)

In order to navigate inside MTP devices it's necessary to:
1) Keep a reference to the root storage folder of the device -> returned by `StorageDevice.FromId()`
2) Keep a reference to the current storage folder
3) Still have a way to retrieve a specific storage folder from a path (necessary e.g when navigating up, when using the NavigationToolbar, etc)

For point (1) I added the property `Root` to the [INavigationControlItem](https://github.com/files-community/files-uwp/pull/92♠6/files#diff-3aa6f76018bffddc35fd94edc3cac47cR10) interface. The property will be populated for removable drives when creating a [DriveItem](https://github.com/files-community/files-uwp/pull/926/files#diff-31a5e6f57b2f992b1b395d718f8a4ddaR41).

For point (2) I added the `CurrentStorageFolder` property to [ItemViewModel](https://github.com/files-community/files-uwp/pull/926/files#diff-c66272bf7eb1f57b0287c4466394719bR43). The `WorkingDirectory` property now simply returns the `CurrentStorageFolder` path. As operations with storage folders are async, changing the `CurrentStorageFolder` is done in a [method](https://github.com/files-community/files-uwp/pull/926/files#diff-c66272bf7eb1f57b0287c4466394719bR54) and not in a property setter.

Point (3) is the real "fun". To get a storage folder from a path:
- I split the given path into components (the same way it's done for NavigationToolbar)
- For each path component, starting from the Root folder (or the CurrentStorageFolder), I drill down calling folder.GetFolderAsync(component) until I get a reference to the storage item I'm interested in.

This is done in [GetFolderFromRelativePathAsync](https://github.com/files-community/files-uwp/pull/926/files#diff-37400bd801d919a8ce26acb7a209f362R179). This behaves the same as `StorageFolder.GetFolderFromPathAsync` unless we are in a removable drive (where the Root property is not null).
This method returns an object that contains the retrieved StorageFolder and a `Path` that in case of removable drives will be a "virtual" path (= "root_name\folder1_name\folder_2_name...")

Most of the other code changes are simply replacing calls to `StorageFolder.GetFolderFromPathAsync` with `StorageFolder.GetFolderFromRelativePathAsync`

Pros:
- no changes for normal drives
- done in UWP (I discarded the idea of doing this in the fulltrust process like the recycle bin)

tsvietOK commented:
@gave92 Get an error when trying to access to my OnePlus 5T file system.
Translated exception: Access denied.
![image](https://user-images.githubusercontent.com/20501502/83977180-335fc100-a907-11ea-8b3a-d389ea736432.png)

gave92 commented:
Thanks for the feedback I only tested this on my phone.
Where inside the RefreshItemsAsync method is that exception thrown?

tsvietOK commented:
In EnumerateItemsFromStandardFolder (ItemViewModel.cs) in foreach:
![image](https://user-images.githubusercontent.com/20501502/83977582-34deb880-a90a-11ea-9d67-f17a7cc97fcb.png)
As i understand, problem is in await _rootFolder.GetFilesAsync()

gave92 commented:
@tsvietOK please let me know if latest commit fixes it thanks!

tsvietOK commented:
@gave92 Ok, now it is working, but much slower then windows explorer. 17sec to load a folder contains 416 elements. 
```
Enumerating items in \\?\ONEPLUS A5010\Внутренний общий накопитель\Music (device) completed in 17396 milliseconds.
```
Now i am getting an exception when trying to create bitmap
![image](https://user-images.githubusercontent.com/20501502/83979045-357c4c80-a914-11ea-81e8-2cb821eba8b2.png)
```
Exception thrown: 'System.UnauthorizedAccessException' in System.Private.CoreLib.dll
An exception of type 'System.UnauthorizedAccessException' occurred in System.Private.CoreLib.dll but was not handled in user code
Access denied. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))
```

tsvietOK commented:
@gave92 Another bug, files are not displayed at all, only folders exist.

gave92 commented:
Ah that's also why you are getting the exception for the bitmap

tsvietOK commented:
Ok, in this case Create file function should be disabled. 
File paste causing app crash. 
If start app when device already connected, app freezes at splash screen.

gave92 commented:
I believe in your case the cause of file paste causing app crash is the same as crash when creating bitmap which is also caused by the UnauthorizedAccessException when enumerating files.
I'm not sure how to debug (and solve) this.
Ok, in this case Create file function should be disabled. -> yup I'll do that as soon as I know this thing doesn't work just on my phone 😂

tsvietOK commented:
Menu "Create file" in windows explorer is not exists for my phone too.
Version from store starting fine with device connected.

gave92 commented:
If you are willing to debug the app and see why it freezes at splash screen it would be a huge help as I can't reproduce it.

tsvietOK commented:
Get another bug. App crash on startup when at least one element pinned to sidebar.
![image](https://user-images.githubusercontent.com/20501502/83981247-e12d9880-a924-11ea-85d3-afdaba7e277c.png)

tsvietOK commented:
Update: App starts fine when device connected, app crash caused by issue above.

gave92 commented:
Good catch should be fixed

yaichenbaum commented:
It looks like CodeFactor found some issues, those will need to be solved either way before merging the pull request, but perhaps they have something to do with the crash.

duke7553 commented:
@gave92 Are we enumerating the storage items 
on the device using the UWP GetItemsAsync() method? Maybe it would be better to somehow get a HANDLE to the StorageFolder and pass it into a Win32 method? It would be a lot faster this was, but I'm not sure if it's possible to get a HANDLE from StorageFolder.

gave92 commented:
@yaichenbaum yeah but they seem to be purely "cosmetic" issue (like multiple classes in the same file..)

@duke7553 yes for mtp devices I'm using GetItemsAsync(UInt32, UInt32) which is slower (it's not so slow on folders so maybe the issue is getting the thumbnail). I don't know if we can use findfirstfileex without a path.

gave92 commented:
@duke7553 Turns out the issue causing very slow folder contents enumeration was getting the item thumbnain. It should be now much faster while still using `GetItemsAsync`.

yaichenbaum commented on Files/Views/ModernShellPage.xaml.cs:
```diff
@@ -152,6 +152,11 @@ private void Page_Loaded(object sender, RoutedEventArgs e)
+                    else if(NavParams.StartsWith("\\\\?\\"))
+                    {
+                        NavigationPath = NavParams;
+                        SidebarControl.SelectedSidebarItem = App.AppSettings.DrivesManager.Drives.First(x => x.Path.ToString().Equals($"{System.IO.Path.GetPathRoot(NavParams)}", StringComparison.OrdinalIgnoreCase));
```
```suggestion
                        SidebarControl.SelectedSidebarItem = AppSettings.DrivesManager.Drives.First(x => x.Path.ToString().Equals($"{System.IO.Path.GetPathRoot(NavParams)}", StringComparison.OrdinalIgnoreCase));
```

gave92 replied:
Apologies this may revert some of the refactoring in #1119 
It's hard to keep this updated :)

yaichenbaum replied:
@gave92 No worries, we are really excited about this pull request 🎉.

tsvietOK commented:
Getting an exception when navigating to my phone storage.
![image](https://user-images.githubusercontent.com/20501502/85641944-b10e2580-b698-11ea-9c1e-f9d2b610590d.png)

gave92 commented:
@tsvietOK thanks for the feedback. The exception should not be thrown on that line. It would be great if you could step through the code and get the line.
On a guess it may be [here](https://github.com/files-community/files-uwp/pull/926/files#diff-c66272bf7eb1f57b0287c4466394719bR112). If that's the case, is `_workingRoot` null at that point?

tsvietOK commented:
@gave92 The problem is in line `return new StorageFolderWithPath(await StorageFolder.GetFolderFromPathAsync(value));` in StorageFileExensions.cs
![image](https://user-images.githubusercontent.com/20501502/85800773-3b629200-b74a-11ea-9e21-cba93aa7a15c.png)
Exception appeared when trying to get storage folder
Value = "\\\\?\\ONEPLUS A5010\\"

gave92 commented:
Is the `rootFolder` parameter null at that point?

tsvietOK commented:
> Is the `rootFolder` parameter of that function `null` when the exception happens?

Yes

gave92 commented on Files/Filesystem/Drives.cs:
```diff
@@ -288,6 +288,29 @@ private void GetVirtualDrivesList(IList<DriveItem> list)
             }
         }
 
+        public static async Task<StorageFolderWithPath> GetRootFromPath(string devicePath)
```
@tsvietOK could you put a breakpoint on this method (GetRootFromPath) and see why it returns `null`? (when you navigate to the phone storage)

tsvietOK replied:
devicePath = "\\\\?\\ONEPLUS A5010\\"
![image](https://user-images.githubusercontent.com/20501502/85863881-70afc400-b7cc-11ea-9e49-73c13c80af82.png)
matchingDrive == null, so this method returns null

gave92 replied:
Thanks a lot. It's not finding a device with matching name so it returns null. Previous method for getting the drive was more robust, guess I'll revert to that.

gave92 replied:
@tsvietOK issue should now be fixed, please let me know.

tsvietOK commented:
Issue fixed, good job! One small glitch: double separator
![image](https://user-images.githubusercontent.com/20501502/85909448-b008ff80-b822-11ea-8aa1-2096169b9b3f.png)

tsvietOK commented on Files/DataModels/SidebarPinnedModel.cs:
```diff
@@ -100,6 +101,18 @@ public async void AddItemToSidebar(string path)
             }
+            catch (Exception ex)
+            {
+                if ((uint)ex.HResult == 0x800700A1)
```
Can you explain?

tsvietOK commented on Files/Interacts/Interaction.cs:
```diff
@@ -433,22 +435,21 @@ private async void OpenSelectedItems(bool displayApplicationPicker)
                     }
-                    else
+                    if(!displayApplicationPicker)
+
```
New line

tsvietOK commented on Files/Views/Pages/Properties.xaml.cs:
```diff
@@ -239,12 +239,12 @@ private async void GetFolderProperties(CancellationTokenSource _tokenSource)
                     GetFolderSize(storageFolder, _tokenSource.Token);
                 }
-            }
+            }            
```
Trailing spaces.

tsvietOK commented on Files/Interacts/Interaction.cs:
```diff
@@ -1095,6 +1108,12 @@ public async Task PasteItems(DataPackageView packageView, string destinationPath
                         // File was moved/deleted in the meantime
                         continue;
                     }
+                    catch (Exception ex)
```
I think there is better way to detect an exception.

gave92 replied:
The thrown exception in case of resource in use is of type "System.Exception".

gave92 commented:
@tsvietOK 
> One small glitch: double separator

Fixed
> And why all pinned(by me) items are placed before default pinned items like Music, Downloads etc.?

That's the magic of not awaiting async methods, should be fixed

tsvietOK commented on Files/DataModels/SidebarPinnedModel.cs:
```diff
@@ -100,14 +102,26 @@ public async void AddItemToSidebar(string path)
             }
+            catch (Exception ex)
+            {
+                if ((uint)ex.HResult == 0x800700A1) // The specified path is invalid (usually an mtp device was disconnected)
```
Could we simplify all exceptions into one?

tsvietOK commented on Files/View Models/ItemViewModel.cs:
```diff
@@ -52,46 +52,89 @@ public class ItemViewModel : INotifyPropertyChanged, IDisposable
-                    _WorkingDirectory = value;
+        public async Task SetWorkingDirectory(string value)
+        {
+            if (string.IsNullOrWhiteSpace(value))
```
```csharp
 if (string.IsNullOrWhiteSpace(value))
{
        return;
}
```

tsvietOK commented:
@gave92 Seems like there is no option to create a folder inside the MTP device.

gave92 commented:
@tsvietOK 
- Right click on empty space > New > Folder should work
- But the menu "+" button on the menu bar is incorrectly disabled
- On right click menu you can also click "New text file/New bitmap" but those do not work.

Do you confirm all this?

tsvietOK commented:
1. ![image](https://user-images.githubusercontent.com/20501502/86047728-057a2200-ba58-11ea-818c-510aa9c8224d.png)
2. "+" button is disabled
3. There is no such menu items at all

tsvietOK commented:
@gave92 Update: Check in grid view mode

gave92 commented:
@tsvietOK thanks I can confirm there is no New options in Grid View Mode

tsvietOK commented:
![image](https://user-images.githubusercontent.com/20501502/86048077-9650fd80-ba58-11ea-81d7-49079c142e61.png)
There is New Bitmap and New Text document menu item exists in context menu in GridViewBrwoser mode

gave92 commented:
@tsvietOK hopefully it's now fixed

yaichenbaum commented:
@gave92 Great work! :shipit:
